### PR TITLE
Test for CountryZoneMember and StateZoneMember (model/schema)

### PR DESCRIPTION
**Pull request by Jeshan7**
## Why?
  - Test cases were less for country and state zone member(model/schema)
  - Coverage was less for country and state zone member(model/schema)

## This change addresses the need by:
 - Wrote new test cases for update_changset in the schema for both country zone
   and state zone member.(Schema)
 - Wrote test cases for get/1, get_all/0, member_changesets/2,
   remove_members_query/2, common_zone_query/2 for both country and
   state zone member.(Model)

[delivers #163514116]

## Checklist

- [ ] I have read [CONTRIBUTING.md][contributing].
- [ ] My code follows the [style guidelines][contributing] of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [ ] I have updated the documentation wherever necessary.
- [ ] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

### CODE REVIEW

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs [hunk 1]
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
     end
   end
 
+  describe "get/1" do
+    test "zone with id", %{zone: zone} do
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/country_zone_test.exs]: Add the failure one too.

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs [hunk 2]
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
...
+      {:ok, new_zone} = CountryZone.get(zone.id)
+      assert new_zone.id == zone.id
+    end
+  end
+
+  describe "get_all/0" do
+    @tag state_zone_count: 3,
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/country_zone_test.exs]: what's the point of setting `state_zone_count`

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs [hunk 3]
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
...
+         country_zone_count: 3
+    test "zones  successfully", %{zones: zone} do
+      country_zones = CountryZone.get_all()
+      cz = Enum.drop_while(country_zones, fn x -> x.zone_type == "C" end)
+      assert cz == []
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/country_zone_test.exs]: you could have easily checked the number of countries set. Please also add test when none are set

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs [hunk 4]
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
...
+    end
+  end
+
+  describe "member_changesets/2" do
+    test "successfully returns a CountryZoneMember changeset", %{zone: zone, countries: countries} do
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/country_zone_test.exs]: I think the statement should be `returns a valid changeset`

Comment (arjun289) [reply apps/snitch_core/test/data/model/zone/country_zone_test.exs]: Also, add scenario when it would be invalid.

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs [hunk 5]
```diff
@@ -77,6 +81,69 @@ defmodule Snitch.Data.Model.CountryZoneTest do
...
+      country_ids = Enum.map(countries, &Map.get(&1, :id))
+      stream = CountryZone.member_changesets(country_ids, zone)
+      changeset = Enum.find_value(stream, fn x -> x end)
+      assert changeset.valid?
+    end
+  end
+
+  describe "remove_members_query/2" do
+    @tag country_zone_count: 3
+    test "returns a query successfully", %{zone: zone, countries: countries} do
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/country_zone_test.exs]: statement should be `returns a valid query`

File: apps/snitch_core/test/data/model/zone/state_zone_test.exs [hunk 6]
```diff
@@ -84,6 +88,69 @@ defmodule Snitch.Data.Model.StateZoneTest do
...
+      assert new_zone.id == zone.id
+    end
+  end
+
+  describe "get_all/0" do
+    @tag state_zone_count: 3,
+         country_zone_count: 3
+    test "zones  successfully", %{zones: zone} do
+      state_zones = StateZone.get_all()
+      sz = Enum.drop_while(state_zones, fn x -> x.zone_type == "S" end)
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/state_zone_test.exs]: Any particular reason you are handling it this way, I am not able to see the purpose.

File: apps/snitch_core/test/data/schema/zone/country_zone_member_test.exs [hunk 7]
```diff
@@ -45,4 +60,39 @@ defmodule Snitch.Data.Schema.CountryZoneMemberTest do
...
+
+      [
+        cs:
+          %CountryZoneMember{}
+          |> CountryZoneMember.create_changeset(params)
+          |> apply_changes()
+      ]
+    end
+
+    test "successfully for valid params", %{cs: cs} do
```
Comment (arjun289) [inline apps/snitch_core/test/data/schema/zone/country_zone_member_test.exs]: should be `returns a valid changeset` instead the statement

**Review by arjun289 — COMMENTED** [review]
I would suggest naming the tests properly. The "describe" along with the test should return a proper statement. Also, consider not using "describe" for functions which have only one test case most likely the `get_all` ones

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs (hunk 2 above)
Comment (Jeshan7) [reply apps/snitch_core/test/data/model/zone/country_zone_test.exs]: To ensure that in the list of structs of country_zone_member and state_zone_member, the get_all/0 function retrieves only those structs which have zone_type set as "C".Similarly in the state_zone_member model also.

File: apps/snitch_core/test/data/model/zone/state_zone_test.exs (hunk 6 above)
Comment (Jeshan7) [reply apps/snitch_core/test/data/model/zone/state_zone_test.exs]: I have used it to access all the struct records inside the list(state_zones). I have used it because only the zone_type field was common to all the records present inside the list(state_zones).

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs (hunk 3 above)
Comment (Jeshan7) [reply apps/snitch_core/test/data/model/zone/country_zone_test.exs]: In this test, I have used zones function present in the factory instead of country_zone function below. get_all/0 retrieves all the zones so while checking I have to assert the count with one extra count added to the country_zone_count tag.  
assert sz(count obtained) == 3 where country_zone_count is set to 2.
I don't think so then it would be an appropriate test because we are changing variables according to our need in the test.

Comment (arjun289) [reply apps/snitch_core/test/data/model/zone/country_zone_test.exs]: `CountryZone.get_all()` returns a list of country zones so you just need to check all the members of this list have zone type as "C". You could have used `Enum.all?(all, fn one -> one.zone_type == "C" end)`

Comment (Jeshan7) [reply apps/snitch_core/test/data/model/zone/country_zone_test.exs]: ok

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs [hunk 8]
```diff
@@ -77,6 +81,85 @@ defmodule Snitch.Data.Model.CountryZoneTest do
...
 
+  describe "get/1" do
+    test "returns a zone using a zone_id", %{zone: zone} do
+      {:ok, new_zone} = CountryZone.get(zone.id)
+      assert new_zone.id == zone.id
+    end
+
+    test "fails for invalid id", %{zone: zone} do
+      {:ok, _} = CountryZone.delete(zone)
+      assert CountryZone.get(zone.id) == {:error, :zone_not_found}
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/country_zone_test.exs]: Instead of deleting the zone you can just pass `-1` as id and it will fail

File: apps/snitch_core/test/data/model/zone/state_zone_test.exs [hunk 9]
```diff
@@ -84,6 +88,85 @@ defmodule Snitch.Data.Model.StateZoneTest do
...
   end
 
+  describe "get/1" do
+    test "returns a state_zone using a zone_id", %{zone: zone} do
+      {:ok, new_zone} = StateZone.get(zone.id)
+      assert new_zone.id == zone.id
+    end
+
+    test "fails for invalid id", %{zone: zone} do
+      {:ok, _} = StateZone.delete(zone)
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/state_zone_test.exs]: see the comment on country `get`

**Review by arjun289 — APPROVED** [review]

File: apps/snitch_core/test/data/model/zone/country_zone_test.exs [hunk 10]
```diff
@@ -3,11 +3,15 @@ defmodule Snitch.Data.Model.CountryZoneTest do
   use Snitch.DataCase
 
   import Snitch.Factory
+  import Ecto.Query
+  alias Ecto.Query
```
Comment (arjun289) [inline apps/snitch_core/test/data/model/zone/country_zone_test.exs]: why both `import` and `alias`? Use any one of them.
